src/widget/textbox.rs

```diff
@@ -35,30 +44,131 @@ const BORDER_WIDTH: f64 = 1.;
 const PADDING_TOP: f64 = 5.;
 const PADDING_LEFT: f64 = 4.;
 
+#[derive(Debug, Clone, Copy)]
+pub struct Selection {
+    /// The inactive edge of a selection, as a byte offset. When
+    /// equal to end, the selection range acts as a caret.
+    pub start: usize,
+
+    /// The active edge of a selection, as a byte offset.
+    pub end: usize,
+}
+
+impl Selection {
+    pub fn new(start: usize, end: usize) -> Self {
+        Selection { start, end }
+    }
+
+    pub fn caret(pos: usize) -> Self {
+        Selection {
+            start: pos,
+            end: pos,
+        }
+    }
+
+    pub fn min(self) -> usize {
+        min(self.start, self.end)
+    }
+
+    pub fn max(self) -> usize {
+        max(self.start, self.end)
+    }
+
+    pub fn is_caret(self) -> bool {
+        self.start == self.end
+    }
+}
+
 #[derive(Debug, Clone)]
 pub struct TextBox {
     width: f64,
+    hscroll_offset: f64,
+    selection: Selection,
+    cursor_timer: TimerToken,
+    cursor_on: bool,
 }
 
 impl TextBox {
     pub fn new(width: f64) -> TextBox {
-        TextBox { width }
+        TextBox {
+            width,
+            hscroll_offset: 0.,
+            selection: Selection::caret(3),
+            cursor_timer: TimerToken::INVALID,
+            cursor_on: false,
+        }
     }
 
-    fn get_layout(
-        &mut self,
-        text: &mut <Piet as RenderContext>::Text,
-        font_size: f64,
-        data: &String,
-    ) -> <Piet as RenderContext>::TextLayout {
+    fn get_layout(&self, text: &mut PietText, font_size: f64, data: &String) -> PietTextLayout {
         // TODO: caching of both the format and the layout
         let font = text
-            .new_font_by_name("Roboto", font_size)
+            .new_font_by_name("Segoe UI", font_size)
             .unwrap()
             .build()
             .unwrap();
         text.new_text_layout(&font, data).unwrap().build().unwrap()
     }
+
+    fn insert(&mut self, src: &mut String, new: &str) {
+        // TODO: handle incomplete graphemes
+        let (left, right) = (self.selection.min(), self.selection.max());
+        src.replace_range(left..right, new);
```

Comment:
TIL this method! It's great, such an improvement over doing truncation and push of slices by hand.